src/Moq/CaptureMatch.cs

```diff
@@ -27,30 +42,80 @@ public CaptureMatch(Action<T> captureCallback)
 		/// <param name="captureCallback">An action to run on captured value</param>
 		/// <param name="predicate">A predicate used to filter captured parameters</param>
 		public CaptureMatch(Action<T> captureCallback, Expression<Func<T, bool>> predicate)
-			: base(CreatePredicate(captureCallback, predicate), () => It.Is(predicate))
+			: this(captureCallback, predicate, false)
+		{
+		}
+		
+		/// <summary>
+		/// Initializes an instance of the capture match.
+		/// </summary>
+		/// <param name="captureCallback">An action to run on captured value</param>
+		/// <param name="predicate">A predicate used to filter captured parameters</param>
+		/// <param name="captureOnSuccessOnly">
+		/// Indicates whether <paramref name="captureCallback"/> should be invoked on parameter evaluation
+		/// or only after all parameters were successfully evaluated.
+		/// </param>
+		public CaptureMatch(Action<T> captureCallback, Expression<Func<T, bool>> predicate, bool captureOnSuccessOnly)
+			: this(new Matcher(captureCallback, predicate, captureOnSuccessOnly))
 		{
 		}
 
-		private static Predicate<T> CreatePredicate(Action<T> captureCallback)
+		private CaptureMatch(Matcher matcher)
+			: base(matcher.Evaluate, matcher.RenderExpression)
 		{
-			return value =>
-			{
-				captureCallback.Invoke(value);
-				return true;
-			};
+			this.matcher = matcher;
 		}
 
-		private static Predicate<T> CreatePredicate(Action<T> captureCallback, Expression<Func<T, bool>> predicate)
+		internal override void OnSuccess()
 		{
-			var predicateDelegate = predicate.CompileUsingExpressionCompiler();
-			return value =>
+			this.matcher.OnSuccess();
+		}
+
+		private class Matcher
+		{
+			private readonly Action<T> captureCallback;
+			private readonly Func<T, bool> predicate;
+			private readonly bool captureOnSuccessOnly;
+			private T capturedValue;
+
+			public Matcher(Action<T> captureCallback, Expression<Func<T, bool>> predicate, bool captureOnSuccessOnly)
+			{
+				this.captureCallback = captureCallback;
+
+				if (predicate != null)
+				{
+					this.predicate = predicate.CompileUsingExpressionCompiler();
+					this.RenderExpression = () => It.Is(predicate);
+				}
+				else
+				{
+					this.predicate = _ => true;
```

Comment:
I'd prefer if we could avoid unnecessary delegate creation. Why not allow `this.predicate` to be `null`, then use null-conditional `this.predicate?.Invoke(...) == true` later?

Reply:
This kind of "static", non-capturing delegate gets cached by the compiler. See https://sharplab.io/#v2:EYLgHgbALAPgAgJgIwFgBQcDMACR2DCAPACoB82A3utjdtbVtgAoBOApgCYCWAxgIYAXNiXL4A9gDtuArpMrYA5mwEBubAGdlagL70ae3DjhRsAWQAUASkoHa2AQAsu6gHTipXGXIC82APrY3uQCLACubCoGumjaQA==